### PR TITLE
Fixes #17897: lost of canonical name of auxiliary constants of mutual fix/cofix at refolding time in "cbn"

### LINKED ISSUE #17897: `cbn` leaves behind unnamable constants when asked to reduce `Pos.add_carry`
#### Description of the problem

```coq
From Coq Require Import PArith.
Local Open Scope positive_scope.
Goal Pos.add 1021 53 = Pos.add 1021 53.
  assert_succeeds (cbv [Pos.add Pos.add_carry Pos.succ]; assert_fails progress cbv).
  assert_succeeds (cbn; assert_fails progress cbv).
  cbn [Pos.add Pos.add_carry Pos.succ]; assert_succeeds progress cbv. (* huh? *)

  lazymatch goal with
  | [ |- (?f 510 26)~0 = (?f 510 26)~0 ]
    => unify f Pos.add_carry;
       idtac f (* Pos.add_carry *)
  end.
  Fail lazymatch goal with
  | [ |- (?f 510 26)~0 = (?f 510 26)~0 ]
    => constr_eq f Pos.add_carry
       end.
  (* The command has indeed failed with message:
Tactic failure: Not equal.
   *)
  Fail progress cbv [Pos.add_carry].
  (* The command has indeed failed with message:
Failed to progress.
   *)
  Set Debug "Cbv".
  cbv [Coq.PArith.BinPos.Pos.add_carry].
  (* Huh? Debug: [Cbv] Not unfolding Coq.PArith.BinPos.Pos.add_carry
Debug: [Cbv] Not unfolding Coq.PArith.BinPos.Pos.add_carry
*)
  cbv. (* Debug: [Cbv] Unfolding Coq.PArith.BinPos.Pos.add_carry
Debug: [Cbv] Unfolding Coq.PArith.BinPos.Pos.succ
*)
```

Presumably this is something to do with mismatched kernel/canonical names?

#### Coq Version

8.17

### PR DESCRIPTION
**Pull request by herbelin**
Fixes / closes #17897

- [x] Added / updated **test-suite**.
- [x] Added **changelog**.

### CODE REVIEW

File: tactics/cbn.ml [hunk 1]
```diff
@@ -569,6 +572,9 @@ let whd_state_gen ?csts flags env sigma =
   let open Context.Named.Declaration in
   let open ReductionBehaviour in
   let rec whrec cst_l (x, stack) =
+    let () = match stack with
```
Comment (ppedrot) [inline tactics/cbn.ml:575]: This looks like remnant of debug code.

Comment (herbelin) [reply tactics/cbn.ml]: Fixed (and changelog at the same time).

In passing, do you know a more compact way than 5 lines to change the label of a constant? Should I make a function in the API of Names for it?

Comment (ppedrot) [reply tactics/cbn.ml]: There is `Names.Constant.change_label`.

Comment (herbelin) [reply tactics/cbn.ml]: Thanks

**Review by ppedrot — APPROVED** [review]

**Comment by ppedrot** [comment]
@coqbot merge now

**Pull request merged**